<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Convenience Functions for Biscuit • biscuiteer</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Convenience Functions for Biscuit">
<meta property="og:description" content="A test harness for bsseq loading of Biscuit output, summarization
             of WGBS data over defined regions and in mappable samples, with or
             without imputation, dropping of mostly-NA rows, age estimates, etc.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">biscuiteer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/biscuiteer.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/trichelab/biscuiteer/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="section" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#section" class="anchor"></a><img src="reference/figures/biscuiteer_logo.png" align="right">
</h1></div>
</div>
<div id="biscuiteer" class="section level1">
<h1 class="hasAnchor">
<a href="#biscuiteer" class="anchor"></a>biscuiteer</h1>

<div id="the-original-luxury-biscuit-boutique" class="section level2">
<h2 class="hasAnchor">
<a href="#the-original-luxury-biscuit-boutique" class="anchor"></a>The original luxury biscuit boutique</h2>
<p>Wait, no, that’s <a href="https://www.biscuiteers.com/">these guys</a>. <code>biscuiteer</code>, on the other hand, is a package to process <a href="https://github.com/zwdzwd/biscuit">biscuit</a> output quickly into <a href="https://bioconductor.org/packages/bsseq">bsseq</a> objects. A number of features such as VCF header parsing, shrunken M-value calculations (for compartment inference), and tumor/normal copy number segmentation are also included, but the task of locus- and region-level differential methylation inference is delegated to other packages (such as <code>dmrseq</code>).</p>
</div>
<div id="installing" class="section level2">
<h2 class="hasAnchor">
<a href="#installing" class="anchor"></a>Installing</h2>
<div class="sourceCode" id="cb1"><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"biscuiteer"</span>)</pre></div>
<p>A development version is available on GitHub and can be installed via:</p>
<div class="sourceCode" id="cb2"><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trichelab/biscuiteerData"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trichelab/biscuiteer"</span>)</pre></div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<div id="loading-data" class="section level3">
<h3 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h3>
<p><code>biscuiteer</code> can load headered or header-free BED-like files as produced from <code>biscuit vcf2bed</code> or <code>biscuit mergecg</code>, but we encourage users to keep their VCF headers (or just the entire VCF, which you will want to do anyways, as biscuit calls SNVs and allows for structural variant detection in a manner similar to typical whole-genome sequencing tools). Since <code>biscuit</code> records the version of the software and the calling arguments used to process a set of files in the output VCF, this allows for much better reproducibility:</p>
<div class="sourceCode" id="cb3"><pre class="r"><span class="co"># Load the package</span>
<span class="co">#</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biscuiteer</span>)

<span class="co"># To read in some data, you need:</span>
<span class="co">#   A BED file (with or without a header)</span>
<span class="co">#   A VCF file (only really needs the header)</span>
<span class="co">#   Whether the file is merged CG data or not</span>
<span class="co">#   Any other additional function inputs (genome, region to load, etc)</span>
<span class="co">#</span>
<span class="no">orig_bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MCF7_Cunha_chr11p15.bed.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">orig_vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MCF7_Cunha_header_only.vcf.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">bisc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/readBiscuit.html">readBiscuit</a></span>(<span class="kw">BEDfile</span> <span class="kw">=</span> <span class="no">orig_bed</span>, <span class="kw">VCFfile</span> <span class="kw">=</span> <span class="no">orig_vcf</span>,
                    <span class="kw">merged</span> <span class="kw">=</span> <span class="fl">FALSE</span>)

<span class="co"># To print metadata information from the loaded file:</span>
<span class="co">#</span>
<span class="fu"><a href="reference/biscuitMetadata.html">biscuitMetadata</a></span>(<span class="no">bisc</span>)
<span class="co">#</span>
<span class="co"># CharacterList of length 3</span>
<span class="co"># [["Reference genome"]] hg19.fa</span>
<span class="co"># [["Biscuit version"]] 0.1.3.20160324</span>
<span class="co"># [["Invocation"]] biscuit pileup -r /primary/vari/genomicdata/genomes/hg19/hg1...</span>

<span class="co"># This is all drawn from the VCF header:</span>
<span class="co">#</span>
<span class="fu">metadata</span>(<span class="no">bisc</span>)$<span class="no">vcfHeader</span>
<span class="co">#</span>
<span class="co"># class: VCFHeader </span>
<span class="co"># samples(1): MCF7_Cunha</span>
<span class="co"># meta(5): fileformat reference source contig program</span>
<span class="co"># fixed(1): FILTER</span>
<span class="co"># info(3): NS CX N5</span>
<span class="co"># geno(7): GT DP ... GL GQ</span></pre></div>
</div>
<div id="downstream-bits" class="section level3">
<h3 class="hasAnchor">
<a href="#downstream-bits" class="anchor"></a>Downstream bits</h3>
<p>A/B compartment inference, age estimation from WGBS, and so forth (examples to appear).<br>
To wit, a shorthand summary of hypo- and hyper-methylation at some regions commonly associated with each:</p>
<div class="sourceCode" id="cb4"><pre class="r"><span class="no">bisc.CpGindex</span> <span class="kw">&lt;-</span> <span class="fu"><a href="reference/CpGindex.html">CpGindex</a></span>(<span class="no">bisc</span>)
<span class="co">#</span>
<span class="co"># Computing hypermethylation indices...</span>
<span class="co"># Loading HMM_CpG_islands.hg19...</span>
<span class="co"># Loading H9state23unmeth.hg19...</span>
<span class="co"># Computing hypomethylation indices...</span>
<span class="co"># Loading PMDs.hg19.rda from biscuiteerData...</span>
<span class="co"># Loading Zhou_solo_WCGW_inCommonPMDs.hg19.rda from biscuiteerData...</span>
<span class="co"># Computing indices...</span>

<span class="fu">show</span>(<span class="no">bisc.CpGindex</span>)
<span class="co">#</span>
<span class="co"># CpGindex with 1 row and 3 columns</span>
<span class="co">#     hyper.MCF7_Cunha   hypo.MCF7_Cunha ratio.MCF7_Cunha</span>
<span class="co">#            &lt;numeric&gt;         &lt;numeric&gt;        &lt;numeric&gt;</span>
<span class="co"># 1 0.0690734126984127 0.199261516805161 0.34664702851757</span>
<span class="co">#   -------</span>
<span class="co"># This object is just a DataFrame that has an idea of where it came from:</span>
<span class="co"># Hypermethylation was tallied across 120 region (see 'object@hyperMethRegions'). </span>
<span class="co"># Hypomethylation was tallied across 13127 region (see 'object@hypoMethRegions').</span>

<span class="no">bisc.CpGindex</span>@<span class="kw">hyperMethRegions</span>
<span class="co">#</span>
<span class="co"># GRanges object with 120 ranges and 1 metadata column:</span>
<span class="co">#       seqnames            ranges strand |              score</span>
<span class="co">#          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |          &lt;numeric&gt;</span>
<span class="co">#     1     chr1 32230201-32230224      * | 0.0399999991059303</span>
<span class="co">#     2     chr1 43638401-43638449      * | 0.0399999991059303</span>
<span class="co">#     3     chr1 44884001-44884005      * | 0.0399999991059303</span>
<span class="co">#     4     chr1 46860401-46860406      * | 0.0599999986588955</span>
<span class="co">#     5     chr1 51435801-51436075      * | 0.0499999998137355</span>
<span class="co">#   ...      ...               ...    ... .                ...</span>
<span class="co">#   116    chr20   8112392-8112400      * | 0.0299999993294477</span>
<span class="co">#   117    chr20 17207801-17208191      * | 0.0599999986588955</span>
<span class="co">#   118    chr22 20004801-20004802      * | 0.0350000001490116</span>
<span class="co">#   119    chr22 37252601-37252731      * | 0.0599999986588955</span>
<span class="co">#   120    chr22 43781850-43781952      * | 0.0449999999254942</span>
<span class="co">#   -------</span>
<span class="co">#   seqinfo: 21 sequences from hg19 genome</span></pre></div>
</div>
<div id="updating-documentation" class="section level3">
<h3 class="hasAnchor">
<a href="#updating-documentation" class="anchor"></a>Updating documentation</h3>
<p><code>$ make doc</code>. Requires the <a href="https://github.com/klutometis/roxygen">roxygen2</a> package.</p>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/biscuiteer">https://​www.bioconductor.org/​packages/​biscuiteer</a>
</li>
<li>Browse source code at <br><a href="https://github.com/trichelab/biscuiteer/">https://​github.com/​trichelab/​biscuiteer/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/trichelab/biscuiteer/issues">https://​github.com/​trichelab/​biscuiteer/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>"Jacob Morrison" <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/ttriche/biscuiteer"><img src="https://travis-ci.org/ttriche/biscuiteer.png?branch=master" alt="Build Status"></a></li>
<li><a href="https://codecov.io/gh/ttriche/biscuiteer"><img src="https://codecov.io/gh/ttriche/biscuiteer/branch/master/graph/badge.svg" alt="codecov"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by "Jacob Morrison".</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
