<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Biscuiteer User Guide • biscuiteer</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Biscuiteer User Guide">
<meta property="og:description" content="biscuiteer">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">biscuiteer</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/biscuiteer.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/trichelab/biscuiteer/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="biscuiteer_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Biscuiteer User Guide</h1>
            
            <h4 class="date">15 July 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/trichelab/biscuiteer/blob/master/vignettes/biscuiteer.Rmd"><code>vignettes/biscuiteer.Rmd</code></a></small>
      <div class="hidden name"><code>biscuiteer.Rmd</code></div>

    </div>

    
    
<div id="biscuiteer" class="section level1">
<h1 class="hasAnchor">
<a href="#biscuiteer" class="anchor"></a>Biscuiteer</h1>
<p><code>biscuiteer</code> is package to process output from <a href="https://github.com/zwdzwd/biscuit">biscuit</a> into <a href="https://bioconductor.org/packages/bsseq">bsseq</a> objects. It includes a number of features, such as VCF header parsing, shrunken M-value calculations (which can be used for compartment inference), and age inference are included. However, the task of locus- and region-level differential methylation inference is delegated to other packages (such as <code>dmrseq</code>).</p>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<div id="installing" class="section level2">
<h2 class="hasAnchor">
<a href="#installing" class="anchor"></a>Installing</h2>
<p>From Bioconductor,</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"biscuiteer"</span>)</pre></body></html></div>
<p>A development version is available on GitHub and can be installed via:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="kw">if</span> (!<span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="kw">quietly</span><span class="kw">=</span><span class="fl">TRUE</span>))
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trichelab/biscuiteerData"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"trichelab/biscuiteer"</span>)</pre></body></html></div>
</div>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading Data</h2>
<p><code>biscuiteer</code> can load either headered of header-free BED files produced from <code>biscuit vcf2bed</code> or <code>biscuit mergecg</code>. In either case, a VCF file is needed when loading <code>biscuit</code> output. For practical purposes, only the VCF header is for <code>biscuiteer</code>. However, it is encouraged that the user keep the entire VCF, as <code>biscuit</code> can be used to call SNVs and allows for structural variant detection in a similar manner to typical whole-genome sequencing tools. Furthermore, <code>biscuit</code> records the version of the software and the calling arguments used during processing the output VCF, which allows for better reproducibility.</p>
<p>NOTE: Both the input BED and VCF files must be tabix’ed before being input to <code>biscuiteer</code>. This can be done by running <code>bgzip biscuit_output.xxx</code> followed by <code>tabix -p xxx biscuit_output.xxx.gz</code>, where <code>xxx</code> is either <code>bed</code> or <code>vcf</code>.</p>
<p>Data can be loaded using the <code>readBiscuit</code> function in <code>biscuiteer</code>:</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biscuiteer</span>)</pre></body></html></div>
<pre><code>## Loading required package: biscuiteerData</code></pre>
<pre><code>## Loading required package: ExperimentHub</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: 'BiocGenerics'</code></pre>
<pre><code>## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Loading required package: AnnotationHub</code></pre>
<pre><code>## Loading required package: BiocFileCache</code></pre>
<pre><code>## Loading required package: dbplyr</code></pre>
<pre><code>## Loading biscuiteerData.</code></pre>
<pre><code>## Loading required package: bsseq</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: 'S4Vectors'</code></pre>
<pre><code>## The following object is masked from 'package:base':
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## 
## Attaching package: 'IRanges'</code></pre>
<pre><code>## The following object is masked from 'package:grDevices':
## 
##     windows</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     'browseVignettes()'. To cite Bioconductor, see
##     'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
<pre><code>## 
## Attaching package: 'Biobase'</code></pre>
<pre><code>## The following object is masked from 'package:ExperimentHub':
## 
##     cache</code></pre>
<pre><code>## The following object is masked from 'package:AnnotationHub':
## 
##     cache</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## 
## Attaching package: 'matrixStats'</code></pre>
<pre><code>## The following objects are masked from 'package:Biobase':
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## 
## Attaching package: 'DelayedArray'</code></pre>
<pre><code>## The following objects are masked from 'package:matrixStats':
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     aperm, apply, rowsum</code></pre>
<pre><code>## </code></pre>
<pre><code>## </code></pre>
<pre><code>## </code></pre>
<pre><code>## Warning: replacing previous import 'BiocParallel::bpstart' by 'QDNAseq::bpstart'
## when loading 'biscuiteer'</code></pre>
<div class="sourceCode" id="cb43"><html><body><pre class="r"><span class="no">orig_bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MCF7_Cunha_chr11p15.bed.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">orig_vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MCF7_Cunha_header_only.vcf.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">bisc</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/readBiscuit.html">readBiscuit</a></span>(<span class="kw">BEDfile</span> <span class="kw">=</span> <span class="no">orig_bed</span>, <span class="kw">VCFfile</span> <span class="kw">=</span> <span class="no">orig_vcf</span>,
                    <span class="kw">merged</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## Checking C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz for import...</code></pre>
<pre><code>## Extracting sample names from C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_header_only.vcf.gz...</code></pre>
<pre><code>## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz does not have a header. Using VCF file header information to help set column names.</code></pre>
<pre><code>## Assuming unmerged data. Checking now... ...The file might be alright. Double check if you're worried.
## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz has 254147 indexed loci.
## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz looks valid for import.
## Reading unmerged input from C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz...
## Excluding CpG sites with uniformly zero coverage...
## Loaded C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15.bed.gz. Creating bsseq object...</code></pre>
<p>Metadata from the <code>biscuit</code> output can be viewed via:</p>
<div class="sourceCode" id="cb48"><html><body><pre class="r"><span class="fu"><a href="../reference/biscuitMetadata.html">biscuitMetadata</a></span>(<span class="no">bisc</span>)</pre></body></html></div>
<pre><code>## CharacterList of length 3
## [["Reference genome"]] hg19.fa
## [["Biscuit version"]] 0.1.3.20160324
## [["Invocation"]] biscuit pileup -r /primary/vari/genomicdata/genomes/hg19/hg1...</code></pre>
<p>If further information about the VCF header is desired,</p>
<div class="sourceCode" id="cb50"><html><body><pre class="r"><span class="fu">metadata</span>(<span class="no">bisc</span>)$<span class="no">vcfHeader</span></pre></body></html></div>
<pre><code>## class: VCFHeader 
## samples(1): MCF7_Cunha
## meta(5): fileformat reference source contig program
## fixed(1): FILTER
## info(3): NS CX N5
## geno(7): GT DP ... GL GQ</code></pre>
</div>
<div id="combining-results" class="section level2">
<h2 class="hasAnchor">
<a href="#combining-results" class="anchor"></a>Combining Results</h2>
<p>In the instance where you have two separate BED files that you would like to analyze in a single bsseq object, you can combine the files using <code>unionize</code>, which is a wrapper around the bsseq function, <code>combine</code>.</p>
<div class="sourceCode" id="cb52"><html><body><pre class="r"><span class="no">shuf_bed</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"MCF7_Cunha_chr11p15_shuffled.bed.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">shuf_vcf</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>,
                        <span class="st">"MCF7_Cunha_shuffled_header_only.vcf.gz"</span>,
                        <span class="kw">package</span><span class="kw">=</span><span class="st">"biscuiteer"</span>)
<span class="no">bisc2</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/readBiscuit.html">readBiscuit</a></span>(<span class="kw">BEDfile</span> <span class="kw">=</span> <span class="no">shuf_bed</span>, <span class="kw">VCFfile</span> <span class="kw">=</span> <span class="no">shuf_vcf</span>,
                     <span class="kw">merged</span> <span class="kw">=</span> <span class="fl">FALSE</span>)</pre></body></html></div>
<pre><code>## Checking C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz for import...</code></pre>
<pre><code>## Extracting sample names from C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_shuffled_header_only.vcf.gz...</code></pre>
<pre><code>## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz does not have a header. Using VCF file header information to help set column names.</code></pre>
<pre><code>## Assuming unmerged data. Checking now... ...The file might be alright. Double check if you're worried.
## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz has 254147 indexed loci.
## C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz looks valid for import.
## Reading unmerged input from C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz...
## Excluding CpG sites with uniformly zero coverage...
## Loaded C:/Users/Hp/R/win-library/4.0/biscuiteer/extdata/MCF7_Cunha_chr11p15_shuffled.bed.gz. Creating bsseq object...</code></pre>
<div class="sourceCode" id="cb57"><html><body><pre class="r"><span class="no">comb</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/unionize.html">unionize</a></span>(<span class="no">bisc</span>, <span class="no">bisc2</span>)</pre></body></html></div>
</div>
</div>
<div id="analysis-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#analysis-functionality" class="anchor"></a>Analysis Functionality</h1>
<p>A handful of analysis paths are available in <code>biscuiteer</code>, including A/B comparment inference, age estimation from WGBS data, hypermethylation of Polycomb Repressor Complex (PRC) binding sites, and hypomethylation of CpG-poor “partially methylated domains” (PMDs).</p>
<div id="inputs-for-ab-compartment-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#inputs-for-ab-compartment-inference" class="anchor"></a>Inputs for A/B Compartment Inference</h2>
<p>When performing A/B compartment inference, the goal is to have something that has roughly gaussian error. <code>getLogitFracMeth</code> uses Dirichlet smoothing to turn raw measurements into lightly moderated, logit-transformed methylated-fraction estimates, which can the be used as inputs to <a href="https://bioconductor.org/packages/release/bioc/html/compartmap.html">compartmap</a></p>
<div class="sourceCode" id="cb58"><html><body><pre class="r"><span class="no">reg</span> <span class="kw">&lt;-</span> <span class="fu">GRanges</span>(<span class="kw">seqnames</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"chr11"</span>,<span class="fl">5</span>),
               <span class="kw">strand</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"*"</span>,<span class="fl">5</span>),
               <span class="kw">ranges</span> <span class="kw">=</span> <span class="fu">IRanges</span>(<span class="kw">start</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>,<span class="fl">2.8e6</span>,<span class="fl">1.17e7</span>,<span class="fl">1.38e7</span>,<span class="fl">1.69e7</span>),
                                <span class="kw">end</span><span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.8e6</span>,<span class="fl">1.17e7</span>,<span class="fl">1.38e7</span>,<span class="fl">1.69e7</span>,<span class="fl">2.2e7</span>))
              )

<span class="no">frac</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getLogitFracMeth.html">getLogitFracMeth</a></span>(<span class="no">bisc</span>, <span class="kw">minSamp</span> <span class="kw">=</span> <span class="fl">1</span>, <span class="kw">r</span> <span class="kw">=</span> <span class="no">reg</span>)
<span class="no">frac</span></pre></body></html></div>
<pre><code>## GRanges object with 5 ranges and 1 metadata column:
##       seqnames            ranges strand | MCF7_Cunha
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;
##   [1]    chr11         0-2800000      * |   1.340682
##   [2]    chr11  2800000-11700000      * |   0.575875
##   [3]    chr11 11700000-13800000      * |   1.162989
##   [4]    chr11 13800000-16900000      * |   0.581874
##   [5]    chr11 16900000-22000000      * |   0.442985
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="age-estimation" class="section level2">
<h2 class="hasAnchor">
<a href="#age-estimation" class="anchor"></a>Age Estimation</h2>
<p><code>biscuiteer</code> has the functionalitity to guess the age of the sample(s) provided using the Horvath-style “clock” models (see <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2013-14-10-r115">Horvath, 2013</a> for more information).</p>
<p>NOTE: The prediction accuracy of this function is entirely dependent on the parameters set by the user. As such, the defaults (as shown in the example below) should only be used as a starting point for exploration by the user.</p>
<p>NOTE: Please cite the appropriate papers for the epigenetic “clock” chosen: * For <code>horvath</code> or <code>horvathshrunk</code> * Horvath, Genome Biology, 2013 * For <code>hannum</code> * Hannum et al., Molecular Cell, 2013 * For <code>skinandblood</code> * Horvath et al., Aging, 2018</p>
<div class="sourceCode" id="cb60"><html><body><pre class="r"><span class="no">ages</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/WGBSage.html">WGBSage</a></span>(<span class="no">comb</span>, <span class="st">"horvath"</span>)</pre></body></html></div>
<pre><code>## Assessing coverage across age-associated regions...</code></pre>
<pre><code>## All regions in all samples appear to be sufficiently covered.</code></pre>
<div class="sourceCode" id="cb63"><html><body><pre class="r"><span class="no">ages</span></pre></body></html></div>
<pre><code>## $call
## WGBSage(comb, "horvath")
## 
## $droppedSamples
## NULL
## 
## $droppedRegions
## NULL
## 
## $intercept
## [1] 0.6955073
## 
## $methcoefs
## GRanges object with 2 ranges and 3 metadata columns:
##                           seqnames            ranges strand | MCF7_Cunha
##                              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;
##     chr11:6678129-6678158    chr11   6678129-6678158      * |   0.800000
##   chr11:12030629-12030658    chr11 12030629-12030658      * |   0.833333
##                           MCF7_Cunha_shuffled        coefs
##                                     &lt;numeric&gt;    &lt;numeric&gt;
##     chr11:6678129-6678158            0.250000  0.000792206
##   chr11:12030629-12030658            0.247732 -0.138857398
##   -------
##   seqinfo: 22 sequences from hg19 genome
## 
## $age
##                         [,1]
## MCF7_Cunha          33.18896
## MCF7_Cunha_shuffled 34.88742</code></pre>
</div>
<div id="hypermethylation-of-prcs-and-hypomethylation-of-pmds" class="section level2">
<h2 class="hasAnchor">
<a href="#hypermethylation-of-prcs-and-hypomethylation-of-pmds" class="anchor"></a>Hypermethylation of PRCs and Hypomethylation of PMDs</h2>
<p>To generate a shorthand summary of the hypermethylation of PRCs and hypomethylation of PMDs, the <code>CpGindex</code> function in <code>biscuiteer</code> can be used.</p>
<div class="sourceCode" id="cb65"><html><body><pre class="r"><span class="no">bisc.CpGindex</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/CpGindex.html">CpGindex</a></span>(<span class="no">bisc</span>)</pre></body></html></div>
<pre><code>## Computing hypermethylation indices...</code></pre>
<pre><code>## Loading HMM_CpG_islands.hg19...</code></pre>
<pre><code>## Loading H9state23unmeth.hg19...</code></pre>
<pre><code>## Computing hypomethylation indices...</code></pre>
<pre><code>## Loading PMDs.hg19.rda from biscuiteerData...</code></pre>
<pre><code>## Loading Zhou_solo_WCGW_inCommonPMDs.hg19.rda from biscuiteerData...</code></pre>
<pre><code>## Computing indices...</code></pre>
<div class="sourceCode" id="cb73"><html><body><pre class="r"><span class="fu">show</span>(<span class="no">bisc.CpGindex</span>)</pre></body></html></div>
<pre><code>## CpGindex with 1 row and 3 columns
##   hyper.MCF7_Cunha hypo.MCF7_Cunha ratio.MCF7_Cunha
##          &lt;numeric&gt;       &lt;numeric&gt;        &lt;numeric&gt;
## 1        0.0690734        0.199262         0.346647
##   -------
## This object is just a DataFrame that has an idea of where it came from:
## Hypermethylation was tallied across 120 region (see 'object@hyperMethRegions'). 
## Hypomethylation was tallied across 13127 region (see 'object@hypoMethRegions').</code></pre>
<div class="sourceCode" id="cb75"><html><body><pre class="r"><span class="no">bisc.CpGindex</span>@<span class="kw">hyperMethRegions</span></pre></body></html></div>
<pre><code>## GRanges object with 120 ranges and 1 metadata column:
##       seqnames            ranges strand |     score
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##     1     chr1 32230201-32230224      * |      0.04
##     2     chr1 43638401-43638449      * |      0.04
##     3     chr1 44884001-44884005      * |      0.04
##     4     chr1 46860401-46860406      * |      0.06
##     5     chr1 51435801-51436075      * |      0.05
##   ...      ...               ...    ... .       ...
##   116    chr20   8112392-8112400      * |     0.030
##   117    chr20 17207801-17208191      * |     0.060
##   118    chr22 20004801-20004802      * |     0.035
##   119    chr22 37252601-37252731      * |     0.060
##   120    chr22 43781850-43781952      * |     0.045
##   -------
##   seqinfo: 21 sequences from hg19 genome</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by "Jacob Morrison".</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
